sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala

```diff
@@ -197,15 +212,24 @@ private[state] class HDFSBackedStateStoreProvider extends StateStoreProvider wit
   }
 
   /** Get the state store for making updates to create a new `version` of the store. */
-  override def getStore(version: Long): StateStore = synchronized {
+  override def getStore(version: Long): StateStore = getStore(version, readOnly = false)
+
+  /** Get the state store for reading to specific `version` of the store. */
+  override def getReadOnlyStore(version: Long): StateStore = getStore(version, readOnly = true)
+
+  private def getStore(version: Long, readOnly: Boolean): StateStore = synchronized {
     require(version >= 0, "Version cannot be less than 0")
     val newMap = new MapType()
     if (version > 0) {
       newMap.putAll(loadMap(version))
     }
-    val store = new HDFSBackedStateStore(version, newMap)
-    logInfo(s"Retrieved version $version of ${HDFSBackedStateStoreProvider.this} for update")
-    store
+    if (readOnly) {
+      logInfo(s"Retrieved version $version of ${HDFSBackedStateStoreProvider.this} for readonly")
+      new HDFSBackedReadOnlyStateStore(version, newMap)
```

Comment:
There might be another chance of optimization here; use the map instance returned from `loadMap(version)` as we always create a new map for `read+write`, but as MapType itself is mutable I would like to avoid any possible bug. 

I've considered changing the type of value in loadedMaps as immutable one to get rid of possibility, but the type we mostly deal with is MapType in any way, and I'm not sure how much overhead will be required during conversion between mutable <-> immutable.